core/src/services/compfs/writer.rs

```diff
@@ -44,27 +48,35 @@ impl oio::Write for CompfsWriter {
     /// the write_all doesn't work correctly if `bs` is non-contiguous.
     ///
     /// The IoBuf::buf_len() only returns the length of the current buffer.
-    async fn write(&mut self, bs: Buffer) -> Result<()> {
-        let mut file = self.file.clone();
+    async fn write(&mut self, mut bs: Buffer) -> Result<()> {
+        let mut file = self.file.clone().unwrap();
 
-        self.core
+        let pos = self
+            .core
             .exec(move || async move {
-                buf_try!(@try file.write_all(bs).await);
-                Ok(())
+                while bs.has_remaining() {
+                    let len = bs.chunk().len();
+                    let res;
+                    BufResult(res, bs) = file.write_all(bs).await;
+                    res?;
+                    bs.advance(len);
+                }
+                Ok(file.position())
             })
             .await?;
+        self.file.as_mut().unwrap().set_position(pos);
 
         Ok(())
     }
 
     async fn close(&mut self) -> Result<Metadata> {
-        let f = self.file.clone();
+        let f = self.file.clone().unwrap();
 
         self.core
             .exec(move || async move { f.get_ref().sync_all().await })
             .await?;
 
-        let f = self.file.clone();
+        let f = self.file.take().unwrap();
```

Comment:
All functions of the writer might re-enter even after returning errors due to retries. 

It's better to use `if let Some(xx) = self.file.take()` instead. We can return errors like `file has closed` in such cases.